src/onnx/parse_attention.cpp

```diff
@@ -0,0 +1,967 @@
+                                 // be equal
+        std::size_t
+            v_hidden_size; // From weights or qkv_hiddn_sizes - differs during cross attention_mode
+        std::size_t
+            sequence_length; // Pulled from input dim(1), must be consistent between other parms
+        std::size_t max_sequence_length;   // Pulled from mask_index
+        std::size_t total_sequence_length; // Pulled from past_seq_length + sequence_length
+        std::size_t query_size;            // Also known as head_size. Derived via num_heads
+        bool has_past_input =
+            false; // Set to true when we have past input. Present output required when set
+        bool has_input_bias = false; // Set to true when we have input_bias
+        bool has_attn_mask  = false; // Set to true when we have an attention mask set
+        bool has_attn_bias  = false; // Set to true when we have an attention bias input
+        bool scale_not_query_sz =
+            false; // Set when a different scale attribute from 1/sqrt(query_size)
+        mask_pad index_pad =
+            mask_pad::no_pad; // Used to track state of input projection mask padding
+        atten_mode attn_type; // Used to determine the attention configuration
+    };
+
+    static void handle_qkv_hidden_size_attr(const onnx_parser& parser,
+                                            const onnx_parser::node_info& info,
+                                            struct attention_attr& attr_out)
+    {
+        auto input_val = parser.parse_value(info.attributes.at("qkv_hidden_sizes"));
+        std::vector<int64_t> qkv_values;
+
+        if(input_val.get_shape().type() != shape::int64_type)
+        {
+            MIGRAPHX_THROW("PARSE_ATTENTION: qkv_hidden_sizes must be int64 type");
+        }
+
+        qkv_values = input_val.get_argument().to_vector<int64_t>();
+
+        if(qkv_values.size() != 3)
+        {
+            MIGRAPHX_THROW("PARSE_ATTENTION: qkv_hidden_sizes must have exactly 3 values");
+        }
+
+        if(qkv_values[0] != qkv_values[1])
+        {
+            MIGRAPHX_THROW("Attention: q and k hidden sizes must be identitical!");
+        }
+
+        std::vector<size_t> qkv_vec{static_cast<size_t>(qkv_values[0]),
+                                    static_cast<size_t>(qkv_values[1]),
+                                    static_cast<size_t>(qkv_values[2])};
+        if(std::any_of(qkv_vec.begin(), qkv_vec.end(), [](auto i) { return (i == 0) or (i < 0); }))
+        {
+            MIGRAPHX_THROW("PARSE_ATTENTION: qkv_hidden_sizes must be nonzero and valid");
+        }
+
+        attr_out.qkv_hidden_sizes = qkv_vec;
+    }
+
+    static void handle_attributes(const onnx_parser& parser,
+                                  const onnx_parser::node_info& info,
+                                  struct attention_attr& attr_out,
+                                  struct attention_inferred& inferred_out)
```

Comment:
Return these as a pair or a tuple instead of using an output parameter. You can assign them with structure bindings when you call the function:

```cpp
auto [parsed_attributes, inferred_attributes] = handle_attributes(parser, info)
```

Reply:
Sure seems like a better way to clean up

Reply:
done, handled the inputs a better way too